crates/subspace-archiving/src/archiver.rs

```diff
@@ -614,56 +631,82 @@ impl Archiver {
         Some(segment)
     }
 
-    /// Take segment as an input, apply necessary transformations and produce archived object mappings.
-    /// Must be called before `produce_archived_segment()`.
-    fn produce_object_mappings(&self, segment: &Segment) -> Vec<GlobalObject> {
+    /// Take the last block continuation in the next segment's items, apply necessary
+    /// transformations, and produce object mappings. Then remove the mappings from that block
+    /// continuation.
+    ///
+    /// Must be called on an incomplete segment.
+    fn produce_next_segment_mappings(&mut self) -> Vec<GlobalObject> {
+        let object_mapping = self.produce_object_mappings(self.buffer.iter());
+
+        // We've already produced these mappings, so we don't need to store them in the buffer.
+        if let Some(SegmentItem::BlockContinuation {
+            bytes: _,
+            object_mapping,
+        }) = self.buffer.back_mut()
+        {
+            object_mapping.objects_mut().clear();
+        }
+
+        object_mapping
+    }
+
+    /// Take the last block in the segment items, apply necessary transformations, and produce
+    /// object mappings.
+    ///
+    /// Must be called in the correct order with `produce_archived_segment()`, because that method
+    /// updates the archiver's state.
+    fn produce_object_mappings<'a>(
+        &self,
+        mut segment_items: impl DoubleEndedIterator<Item = &'a SegmentItem>,
```

Comment:
Not sure what is the benefit of the `impl DoubleEndedIterator` for internal function, you could have just done `let segment_items = segment_items.iter()` locally and get the same result.

Reply:
This method is called with `segment.items()` for Block, BlockStart, and a BlockContinuation that fills a segment, and `self.buffer.items()` for the final BlockContinuation which doesn’t fill a segment.

Reply:
Can still be `&[SegmentItem]` or rather `&mut [SegmentItem]` if draining object mappings.